File: backend/users_n_groups/idp.py
from aws_cdk import CfnOutput, Duration, RemovalPolicy, Stack
from aws_cdk import aws_cloudfront as cloudfront
from aws_cdk import aws_cognito as cognito
from aws_cdk import aws_iam as iam
from cdk_nag import NagSuppressions
from constructs import Construct

from backend.users_n_groups.user_pool_user import UserPoolUser


class Idp(Construct):
    @property
    def user_pool(self) -> cognito.IUserPool:
        return self._user_pool

    @property
    def identity_pool(self) -> cognito.CfnIdentityPool:
        return self._identity_pool

    @property
    def admin_user_role(self) -> iam.IRole:
        return self._admin_user_role

    @property
    def operator_user_role(self) -> iam.IRole:
        return self._operator_user_role

    @property
    def unauthenticated_user_role(self) -> iam.IRole:
        return self._unauth_user_role

    def __init__(
        self,
        scope: Construct,
        id: str,
        distribution: cloudfront.IDistribution,
        default_admin_email: str,
        **kwargs,
    ):
        super().__init__(scope, id, **kwargs)

        stack = Stack.of(self)

        self._user_pool = cognito.UserPool(
            self,
            "UserPool",
            user_pool_name=stack.stack_name,
            standard_attributes=cognito.StandardAttributes(
                email=cognito.StandardAttribute(required=True, mutable=True)
            ),
            mfa=cognito.Mfa.OFF,
            self_sign_up_enabled=True,
            auto_verify=cognito.AutoVerifiedAttrs(email=True),
            removal_policy=RemovalPolicy.DESTROY,
            password_policy=cognito.PasswordPolicy(
                min_length=8,
                require_lowercase=True,
                require_uppercase=True,
                require_digits=True,
                require_symbols=True,
                temp_password_validity=Duration.days(2),
            ),
            user_invitation=cognito.UserInvitationConfig(
                email_subject="Invite to join DREM",
                email_body=(
                    "Hello {username}, you have been invited to join DREM. \nYour"
                    " temporary password is \n\n{####}\n\n"
                )
                + "https://"
                + distribution.distribution_domain_name,
                sms_message=(
                    "Hello {username}, your temporary password for DREM is {####}"
                ),
            ),
            user_verification=cognito.UserVerificationConfig(
                email_subject="Verify your email for DREM",
                email_body=(
                    "Thanks for signing up to DREM \n\nYour verification code is"
                    " \n{####}"
                ),
                email_style=cognito.VerificationEmailStyle.CODE,
                sms_message=(
                    "Thanks for signing up to DREM. Your verification code is {####}"
                ),
            ),
        )

        NagSuppressions.add_resource_suppressions(
            self._user_pool,
            suppressions=[
                {
                    "id": "AwsSolutions-COG2",
                    "reason": (
                        "users only sign up and us DREM for a short period of time, all"
                        " users are deleted after 10 days inactivity"
                    ),
                },
                {
                    "id": "AwsSolutions-COG3",
                    "reason": (
                        "users only sign up and us DREM for a short period of time, all"
                        " users are deleted after 10 days inactivity"
                    ),
                },
            ],
        )

        # Cognito Client
        user_pool_client_web = cognito.UserPoolClient(
            self,
            "UserPoolClientWeb",
            user_pool=self._user_pool,
            prevent_user_existence_errors=True,
        )

        cfn_user_pool_client_web = user_pool_client_web.node.default_child
        cfn_user_pool_client_web.callback_ur_ls = [
            "https://" + distribution.distribution_domain_name,
            "http://localhost:3000",
        ]
        cfn_user_pool_client_web.logout_ur_ls = [
            "https://" + distribution.distribution_domain_name,
            "http://localhost:3000",
        ]

        # Cognito Identity Pool
        self._identity_pool = cognito.CfnIdentityPool(
            self,
            "IdentityPool",
            allow_unauthenticated_identities=False,
            cognito_identity_providers=[
                cognito.CfnIdentityPool.CognitoIdentityProviderProperty(
                    client_id=user_pool_client_web.user_pool_client_id,
                    provider_name=self._user_pool.user_pool_provider_name,
                )
            ],
        )

        # Cognito Identity Pool Authenitcated Role
        id_pool_auth_user_role = iam.Role(
            self,
            "CognitoDefaultAuthenticatedRole",
            assumed_by=iam.FederatedPrincipal(
                federated="cognito-identity.amazonaws.com",
                conditions={
                    "StringEquals": {
                        "cognito-identity.amazonaws.com:aud": self._identity_pool.ref,
                    },
                    "ForAnyValue:StringLike": {
                        "cognito-identity.amazonaws.com:amr": "authenticated",
                    },
                },
                assume_role_action="sts:AssumeRoleWithWebIdentity",
            ),
        )

        # #Cognito Identity Pool Unauthenitcated Role
        # needed for accessing stream overlays
        self._unauth_user_role = iam.Role(
            self,
            "CognitoDefaultUnauthenticatedRole",
            assumed_by=iam.FederatedPrincipal(
                federated="cognito-identity.amazonaws.com",
                conditions={
                    "StringEquals": {
                        "cognito-identity.amazonaws.com:aud": self._identity_pool.ref,
                    },
                    "ForAnyValue:StringLike": {
                        "cognito-identity.amazonaws.com:amr": "unauthenticated",
                    },
                },
                assume_role_action="sts:AssumeRoleWithWebIdentity",
            ),
        )

        cognito.CfnIdentityPoolRoleAttachment(
            self,
            "IdentityPoolRoleAttachment",
            identity_pool_id=self._identity_pool.ref,
            roles={
                "authenticated": id_pool_auth_user_role.role_arn,
                "unauthenticated": self._unauth_user_role.role_arn,
            },
            role_mappings={
                "role_mapping": cognito.CfnIdentityPoolRoleAttachment.RoleMappingProperty(  # noqa: E501
                    type="Token",
                    identity_provider="{}:{}".format(
                        self._user_pool.user_pool_provider_name,
                        user_pool_client_web.user_pool_client_id,
                    ),
                    ambiguous_role_resolution="AuthenticatedRole",
                )
            },
        )

        # Admin Users Group Role
        self._admin_user_role = iam.Role(
            self,
            "AdminUserRole",
            assumed_by=iam.FederatedPrincipal(
                federated="cognito-identity.amazonaws.com",
                conditions={
                    "StringEquals": {
                        "cognito-identity.amazonaws.com:aud": self._identity_pool.ref,
                    },
                    "ForAnyValue:StringLike": {
                        "cognito-identity.amazonaws.com:amr": "authenticated",
                    },
                },
                assume_role_action="sts:AssumeRoleWithWebIdentity",
            ),
        )

        # Operator Users Group Role
        self._operator_user_role = iam.Role(
            self,
            "OperatorUserRole",
            assumed_by=iam.FederatedPrincipal(
                federated="cognito-identity.amazonaws.com",
                conditions={
                    "StringEquals": {
                        "cognito-identity.amazonaws.com:aud": self._identity_pool.ref,
                    },
                    "ForAnyValue:StringLike": {
                        "cognito-identity.amazonaws.com:amr": "authenticated",
                    },
                },
                assume_role_action="sts:AssumeRoleWithWebIdentity",
            ),
        )

        # Cognito User Group (Operator)
        cognito.CfnUserPoolGroup(
            self,
            "OperatorGroup",
            user_pool_id=self._user_pool.user_pool_id,
            description="Operator user group",
            group_name="operator",
            role_arn=self._operator_user_role.role_arn,
            precedence=1,
        )

        # Cognito User Group (Admin)
        admin_user_pool_group = cognito.CfnUserPoolGroup(
            self,
            "AdminGroup",
            user_pool_id=self._user_pool.user_pool_id,
            description="Admin user group",
            group_name="admin",
            role_arn=self._admin_user_role.role_arn,
            precedence=1,
        )

        # Add a default Admin user to the system
        default_admin_user_name = "admin"

        UserPoolUser(
            self,
            "DefaultAdminUser",
            username=default_admin_user_name,
            email=default_admin_email,
            user_pool=self._user_pool,
            group_name=admin_user_pool_group.ref,
        )

        # Outputs
        self.userPoolWebClientId = CfnOutput(
            self, "userPoolWebClientId", value=user_pool_client_web.user_pool_client_id
        )

        self.identityPoolId = CfnOutput(
            self, "identityPoolId", value=self._identity_pool.ref
        )

        self.userPoolId = CfnOutput(
            self, "userPoolId", value=self._user_pool.user_pool_id
        )

        CfnOutput(self, "DefaultAdminUserUsername", value=default_admin_user_name)

        CfnOutput(self, "DefaultAdminEmail", value=default_admin_email)

File: lib/constructs/idp.ts
import * as cdk from 'aws-cdk-lib';
import { DockerImage, Duration, RemovalPolicy } from 'aws-cdk-lib';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import * as cognito from 'aws-cdk-lib/aws-cognito';
import { CfnUserPoolUserToGroupAttachment } from 'aws-cdk-lib/aws-cognito';
import { EventBus } from 'aws-cdk-lib/aws-events';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';
import { StandardLambdaPythonFunction } from './standard-lambda-python-function';

/** Group names in the external Cognito User Pool that map to DREM roles.
 *  Only relevant when `existingUserPoolId` is set.
 *  Defaults match the DROA reference group names.
 */
export interface ExternalGroupNames {
  admin?: string;
  operator?: string;
  racer?: string;
  commentator?: string;
  registration?: string;
}

export interface IdpProps {
  distribution: cloudfront.IDistribution;
  defaultAdminEmail: string;
  lambdaConfig: {
    runtime: lambda.Runtime;
    architecture: lambda.Architecture;
    bundlingImage: DockerImage;
    layersConfig: {
      powerToolsLogLevel: string;
      helperFunctionsLayer: lambda.ILayerVersion;
      powerToolsLayer: lambda.ILayerVersion;
    };
  };
  eventbus: EventBus;
  existingUserPoolId?: string;
  /** Override the external pool group names that map to DREM roles (only used when `existingUserPoolId` is set). */
  externalGroupNames?: ExternalGroupNames;
}

export class Idp extends Construct {
  public readonly userPool: cognito.IUserPool;
  public readonly identityPool: cognito.CfnIdentityPool;
  public readonly userPoolClientWeb: cognito.UserPoolClient;
  public readonly authenticatedUserRole: iam.IRole;
  public readonly adminGroupRole: iam.IRole;
  public readonly operatorGroupRole: iam.IRole;
  public readonly commentatorGroupRole: iam.IRole;
  public readonly registrationGroupRole: iam.IRole;
  public readonly unauthenticated_user_role: iam.IRole;
  public readonly useExistingUserPool: boolean;

  constructor(scope: Construct, id: string, props: IdpProps) {
    super(scope, id);

    const stack = cdk.Stack.of(this);

    const useExistingUserPool = !!props.existingUserPoolId;
    this.useExistingUserPool = useExistingUserPool;

    let userPool: cognito.IUserPool;
    let pre_token_generation_lambda: StandardLambdaPythonFunction | undefined;

    if (!useExistingUserPool) {
      const pre_sign_up_lambda = new StandardLambdaPythonFunction(this, 'pre_sign_up_lambda', {
        entry: 'lib/lambdas/cognito_triggers/',
        description: 'Cognito pre sign up Lambda',
        index: 'index.py',
        handler: 'pre_sign_up_handler',
        timeout: Duration.minutes(1),
        runtime: props.lambdaConfig.runtime,
        memorySize: 128,
        architecture: props.lambdaConfig.architecture,
        environment: {
          eventbus_name: props.eventbus.eventBusName,
          default_user_group: 'racer',
          POWERTOOLS_SERVICE_NAME: 'cognito_presignup_lambda',
        },
        bundling: {
          image: props.lambdaConfig.bundlingImage,
        },
        layers: [props.lambdaConfig.layersConfig.helperFunctionsLayer, props.lambdaConfig.layersConfig.powerToolsLayer],
      });
      props.eventbus.grantPutEventsTo(pre_sign_up_lambda);

      const localPreTokenLambda = new StandardLambdaPythonFunction(this, 'post_authentication_lambda', {
        entry: 'lib/lambdas/cognito_triggers/',
        description: 'Cognito pre token trigger lambda',
        handler: 'pre_token_generation_handler',
        runtime: props.lambdaConfig.runtime,
        architecture: props.lambdaConfig.architecture,
        environment: {
          eventbus_name: props.eventbus.eventBusName,
          default_user_group: 'racer',
          POWERTOOLS_SERVICE_NAME: 'cognito_post_confirmation_lambda',
        },
        bundling: {
          image: props.lambdaConfig.bundlingImage,
        },
        layers: [props.lambdaConfig.layersConfig.helperFunctionsLayer, props.lambdaConfig.layersConfig.powerToolsLayer],
      });
      pre_token_generation_lambda = localPreTokenLambda;

      userPool = new cognito.UserPool(this, 'UserPool', {
        userPoolName: stack.stackName,
        signInAliases: { email: true },
        standardAttributes: {
          email: { required: true, mutable: true },
          preferredUsername: { required: false, mutable: true },
        },
        customAttributes: {
          countryCode: new cognito.StringAttribute({
            mutable: true,
          }),
          racerName: new cognito.StringAttribute({
            mutable: true,
          }),
        },
        featurePlan: cognito.FeaturePlan.PLUS,
        mfa: cognito.Mfa.OPTIONAL,
        advancedSecurityMode: cognito.AdvancedSecurityMode.ENFORCED,
        selfSignUpEnabled: true,
        autoVerify: { email: true },
        removalPolicy: RemovalPolicy.DESTROY,
        passwordPolicy: {
          minLength: 8,
          requireLowercase: true,
          requireDigits: true,
          requireSymbols: true,
          requireUppercase: true,
          tempPasswordValidity: Duration.days(2),
        },
        userInvitation: {
          emailSubject: 'Invite to join DREM',
          emailBody:
            'Hello {username}, you have been invited to join DREM. \n' +
            'Your temporary password is \n\n{####}\n\n' +
            'https://' +
            props.distribution.distributionDomainName,
          smsMessage: 'Hello {username}, your temporary password for DREM is {####}',
        },
        userVerification: {
          emailSubject: 'Verify your email for DREM',
          emailBody: 'Thanks for signing up to DREM \n\nYour verification code is \n{####}',
          emailStyle: cognito.VerificationEmailStyle.CODE,
          smsMessage: 'Thanks for signing up to DREM. Your verification code is {####}',
        },
        lambdaTriggers: {
          preSignUp: pre_sign_up_lambda,
          preTokenGeneration: localPreTokenLambda,
        },
      });
    } else {
      userPool = cognito.UserPool.fromUserPoolId(this, 'UserPool', props.existingUserPoolId!);
    }

    this.userPool = userPool;

    if (pre_token_generation_lambda) {
      pre_token_generation_lambda.addAdditionalRolePolicy(
        'updateUserPolicy',
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['cognito-idp:AdminAddUserToGroup'],
          resources: [userPool.userPoolArn],
        })
      );
    }

    //         NagSuppressions.add_resource_suppressions(
    //             self._userPool,
    //             suppressions=[
    //                 {
    //                     "id": "AwsSolutions-COG2",
    //                     "reason": (
    //                         "users only sign up and us DREM for a short period of time, all"
    //                         " users are deleted after 10 days inactivity"
    //                     ),
    //                 },
    //                 {
    //                     "id": "AwsSolutions-COG3",
    //                     "reason": (
    //                         "users only sign up and us DREM for a short period of time, all"
    //                         " users are deleted after 10 days inactivity"
    //                     ),
    //                 },
    //             ],
    //         )

    //  Cognito Client
    const userPoolClientWeb = new cognito.UserPoolClient(this, 'UserPoolClientWeb', {
      userPool: userPool,
      preventUserExistenceErrors: true,
    });

    userPool.addClient('DremClient', {
      oAuth: {
        callbackUrls: ['https://' + props.distribution.distributionDomainName, 'http://localhost:3000'],
        logoutUrls: ['https://' + props.distribution.distributionDomainName, 'http://localhost:3000'],
      },
    });
    this.userPoolClientWeb = userPoolClientWeb;

    //  Cognito Identity Pool
    const identityPool = new cognito.CfnIdentityPool(this, 'IdentityPool', {
      allowUnauthenticatedIdentities: false,
      cognitoIdentityProviders: [
        {
          clientId: userPoolClientWeb.userPoolClientId,
          providerName: userPool.userPoolProviderName,
        },
      ],
    });

    this.identityPool = identityPool;

    // Cognito Identity Pool Authenticated Role (used by the racer group)
    const authUserRole = new iam.Role(this, 'CognitoDefaultAuthenticatedRole', {
      assumedBy: new iam.FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: { 'cognito-identity.amazonaws.com:aud': identityPool.ref },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'authenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
    });
    this.authenticatedUserRole = authUserRole;

    let defaultUserPoolRoleParameter: string | undefined;
    if (pre_token_generation_lambda) {
      defaultUserPoolRoleParameter = `/${stack.stackName}/defaultUserRole`;
      const defaultUserPoolRoleParameterSSM = new ssm.StringParameter(this, 'appsyncHelperLambdaLayerArn', {
        stringValue: authUserRole.roleArn,
        parameterName: defaultUserPoolRoleParameter,
      });
      pre_token_generation_lambda.addAdditionalRolePolicy(
        'getSSMDefaultRoleARNPermission',
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['ssm:GetParameter'],
          resources: [defaultUserPoolRoleParameterSSM.parameterArn],
        })
      );
    }

    //  Admin Users Group Role
    const adminGroupRole = new iam.Role(this, 'AdminUserGroupRole', {
      assumedBy: new iam.FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: { 'cognito-identity.amazonaws.com:aud': identityPool.ref },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'authenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
    });
    this.adminGroupRole = adminGroupRole;

    // Operator Users Group Role
    const operatorGroupRole = new iam.Role(this, 'OperatorUserRole', {
      assumedBy: new iam.FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: { 'cognito-identity.amazonaws.com:aud': identityPool.ref },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'authenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
    });
    this.operatorGroupRole = operatorGroupRole;

    // Commentator Users Group Role
    const commentatorGroupRole = new iam.Role(this, 'CommentatorUserRole', {
      assumedBy: new iam.FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: { 'cognito-identity.amazonaws.com:aud': identityPool.ref },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'authenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
    });
    this.commentatorGroupRole = commentatorGroupRole;

    // Registration Users Group Role
    const registrationGroupRole = new iam.Role(this, 'RegistrationGroupRole', {
      assumedBy: new iam.FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: { 'cognito-identity.amazonaws.com:aud': identityPool.ref },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'authenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
    });
    this.registrationGroupRole = registrationGroupRole;

    // Only resolved for the external pool path (Rules-based mapping).
    const extGroups = useExistingUserPool
      ? {
          admin: props.externalGroupNames?.admin ?? 'dr-admins',
          operator: props.externalGroupNames?.operator ?? 'dr-race-facilitators',
          racer: props.externalGroupNames?.racer ?? 'dr-racers',
          commentator: props.externalGroupNames?.commentator ?? 'dr-commentator',
          registration: props.externalGroupNames?.registration ?? 'dr-registration',
        }
      : undefined;

    new cognito.CfnIdentityPoolRoleAttachment(this, 'IdentityPoolRoleAttachment', {
      identityPoolId: identityPool.ref,
      roles: {
        authenticated: authUserRole.roleArn,
      },
      roleMappings: useExistingUserPool
        ? {
            // Rules-based: map external IDP group names directly to DREM IAM roles
            mapping: {
              type: 'Rules',
              identityProvider: `cognito-idp.${stack.region}.amazonaws.com/${userPool.userPoolId}:${userPoolClientWeb.userPoolClientId}`,
              ambiguousRoleResolution: 'AuthenticatedRole',
              rulesConfiguration: {
                rules: [
                  {
                    claim: 'cognito:groups',
                    matchType: 'Contains',
                    value: extGroups!.admin,
                    roleArn: adminGroupRole.roleArn,
                  },
                  {
                    claim: 'cognito:groups',
                    matchType: 'Contains',
                    value: extGroups!.operator,
                    roleArn: operatorGroupRole.roleArn,
                  },
                  {
                    claim: 'cognito:groups',
                    matchType: 'Contains',
                    value: extGroups!.racer,
                    roleArn: authUserRole.roleArn,
                  },
                  {
                    claim: 'cognito:groups',
                    matchType: 'Contains',
                    value: extGroups!.commentator,
                    roleArn: commentatorGroupRole.roleArn,
                  },
                  {
                    claim: 'cognito:groups',
                    matchType: 'Contains',
                    value: extGroups!.registration,
                    roleArn: registrationGroupRole.roleArn,
                  },
                ],
              },
            },
          }
        : {
            // Token-based: derives roles from Cognito group roleArn assignments
            mapping: {
              type: 'Token',
              identityProvider: `cognito-idp.${stack.region}.amazonaws.com/${userPool.userPoolId}:${userPoolClientWeb.userPoolClientId}`,
              ambiguousRoleResolution: 'AuthenticatedRole',
            },
          },
    });

    if (pre_token_generation_lambda && defaultUserPoolRoleParameter) {
      pre_token_generation_lambda.addEnvironment('default_user_group_role_parameter', defaultUserPoolRoleParameter);
    }

    if (!useExistingUserPool) {
      //  Cognito User Group (Racers)
      new cognito.CfnUserPoolGroup(this, 'RacerGroup', {
        userPoolId: userPool.userPoolId,
        description: 'Racer user group',
        groupName: 'racer',
        precedence: 1,
        roleArn: authUserRole.roleArn,
      });

      //  Cognito User Group (Operator)
      new cognito.CfnUserPoolGroup(this, 'OperatorGroup', {
        userPoolId: userPool.userPoolId,
        description: 'Operator user group',
        groupName: 'operator',
        precedence: 1,
        roleArn: operatorGroupRole.roleArn,
      });

      //  Cognito User Group (Admin)
      const adminGroup = new cognito.CfnUserPoolGroup(this, 'AdminGroup', {
        userPoolId: userPool.userPoolId,
        description: 'Admin user group',
        groupName: 'admin',
        precedence: 1,
        roleArn: adminGroupRole.roleArn,
      });

      //  Add a default Admin user to the system
      const adminUser = new CognitoUser(this, 'DefaultAdminUser', {
        username: 'admin',
        email: props.defaultAdminEmail,
        userPool: userPool,
        groupName: adminGroup.ref,
      });
      adminUser.node.addDependency(adminGroup);
      adminUser.node.addDependency(userPool);

      //  Cognito User Group (Commentator)
      new cognito.CfnUserPoolGroup(this, 'CommentatorGroup', {
        userPoolId: userPool.userPoolId,
        description: 'Commentator user group',
        groupName: 'commentator',
        precedence: 1,
        roleArn: commentatorGroupRole.roleArn,
      });

      //  Cognito User Group (Registration)
      new cognito.CfnUserPoolGroup(this, 'RegistrationGroup', {
        userPoolId: userPool.userPoolId,
        description: 'Registration user group',
        groupName: 'registration',
        precedence: 1,
        roleArn: registrationGroupRole.roleArn,
      });
    }

    NagSuppressions.addResourceSuppressions(
      userPool,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason:
            'The smsRole generated by CDK for Cognito MFA/SMS uses Resource::* on sns:Publish; this role is auto-created by CDK and cannot be scoped to a specific resource.',
          appliesTo: ['Resource::*'],
        },
      ],
      true
    );
  }
}

export interface CognitoUserProps {
  username: string;
  email: string;
  userPool: cognito.IUserPool;
  groupName?: string;
}

export class CognitoUser extends Construct {
  // Creates a user in the provided Cognito User pool

  constructor(scope: Construct, id: string, props: CognitoUserProps) {
    super(scope, id);

    const user = new cognito.CfnUserPoolUser(this, 'adminUser', {
      userPoolId: props.userPool.userPoolId,
      username: props.username,
      desiredDeliveryMediums: ['EMAIL'],
      userAttributes: [{ name: 'email', value: props.email }],
    });

    // If a Group Name is provided, also add the user to this Cognito UserPool Group
    if (props.groupName) {
      const userToGroupAttachment = new CfnUserPoolUserToGroupAttachment(this, 'AdminUserToAdminGroup', {
        userPoolId: user.userPoolId,
        groupName: props.groupName,
        username: user.username!,
      });

      userToGroupAttachment.node.addDependency(user);
      userToGroupAttachment.node.addDependency(props.userPool);
    }
  }
}
